File: BHOI_/2014/BHOI 2014/Transformacija/grader.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

// Jednostavan generator pseudo slucajnih brojeva
unsigned long seed = 123;
inline unsigned myrand() { 
  seed = (214013 * seed + 2531011);
  return (seed>>16)&0x7FFF;
}

// Vraca dvostruki srednji razmak k brojeva na segmentu [0,n] zaokruzen na stepen
// broja dva umanjen za jedan.
inline unsigned IzracunajMaxDeltu(int n, int k) {
	unsigned delta = round(n / k);
	delta |= delta >> 1;
	delta |= delta >> 2;
	delta |= delta >> 4;
	delta |= delta >> 8;
	delta |= delta >> 16;
	delta = delta == 0 ? 1 : delta;
	return delta;
}

int TransformisiNiz(int n, int* krajnji, int m, int* pocetni, int* transformacija);

int GenerisiSortiranNiz(int* arr, int n, int k) {
	unsigned max_delta = IzracunajMaxDeltu(n, k);
	arr[0] = (myrand() & max_delta) % (n - 1) + 1;
	int i;
	for (i = 1; i < k; i++) {
		arr[i] = arr[i - 1] + (myrand() & max_delta);
		if (arr[i] >= n) {
			return i;
		}
	}
	return k;
}

int main() {
	int seed, n, m, l, dummy;
	dummy = scanf("%u %d %d %d", &seed, &n, &m, &l);
	int *krajnji = (int*) malloc(n * sizeof(int)), *pocetni = (int*) malloc(m * sizeof(int)), *transformacija = (int*) malloc((n + m) * sizeof(int));
	n = GenerisiSortiranNiz(krajnji, l, n);
	m = GenerisiSortiranNiz(pocetni, l, m);
	int *krajnji_kopija = (int*) malloc(n * sizeof(int)), *pocetni_kopija = (int*) malloc(m * sizeof(int));
	memcpy(krajnji_kopija, krajnji, n * sizeof(int));
	memcpy(pocetni_kopija, pocetni, m * sizeof(int));
	int r = TransformisiNiz(n, krajnji_kopija, m, pocetni_kopija, transformacija);
	int i = 0, j = 0, k;
	for (k = 0; k < r; k++) {
		if (transformacija[k] == -1) {
			++j;
		} else if (transformacija[k] == 0) {
			if (i >= n || j >= m || krajnji[i] != pocetni[j]) {
				printf("Neispravna transformacija.\n");
				return 0;
			}
			++i; ++j;
		} else {
			if (i >= n || krajnji[i] != transformacija[k]) {
				printf("Neispravna transformacija.\n");
				return 0;
			}
			++i;
		}
	}
	if (i != n || j != m) {
		printf("Neispravna transformacija.\n");
	} else {
		printf("Ispravna transformacija u %d koraka.\n", r);
	}
	return dummy != dummy;
}

File: BHOI_/2014/BHOI 2014/Transformacija/grader.pas
Program grader;
uses transformacija;

var
  seed : LongInt;
  
function myrand() : LongInt;
begin
	seed := (214013 * seed + 2531011);
	myrand := (seed shr 16) and $7FFF;
end;

function IzracunajMaxDeltu(n, k : LongInt) : LongInt;
var
	delta : LongInt;
begin
	delta := round(n / k);
	delta := delta or (delta shr 1);
	delta := delta or (delta shr 2);
	delta := delta or (delta shr 4);
	delta := delta or (delta shr 8);
	delta := delta or (delta shr 16);
	if (delta = 0) then delta := 1;
	IzracunajMaxDeltu := delta;
end;

function GenerisiSortiranNiz(var arr : Array of LongInt; n, k : LongInt) : LongInt;
var
	max_delta, i : LongInt;
begin
	max_delta := IzracunajMaxDeltu(n, k);
	arr[0] := ((myrand() and max_delta) mod (n - 1)) + 1;
	for i := 1 to k-1 do
		begin
		arr[i] := arr[i - 1] + (myrand() and max_delta);
		if (arr[i] >= n) then
			begin
			Exit(i);
			end;
		end;
	Exit(k);
end;

var
  n, m, l, r, i, j, k : LongInt;
  krajnji, pocetni, rezultat : Array of LongInt;

begin
  read(seed, n, m, l);
  SetLength(krajnji, n);
  n := GenerisiSortiranNiz(krajnji, l, n);
  SetLength(pocetni, m);
  m := GenerisiSortiranNiz(pocetni, l, m);
  SetLength(rezultat, m + n);
  r := TransformisiNiz(n, Copy(krajnji, 0, n), m, Copy(pocetni, 0, m), rezultat);
  i := 0; j:= 0;
	for k := 0 to r - 1 do
		begin
		if rezultat[k] = -1 then
			j := j + 1
		else if rezultat[k] = 0 then
			begin
			if (i >= n) or (j >= m) or (krajnji[i] <> pocetni[j]) then
				begin
				writeln('Neispravna transformacija');
				Exit;
				end;
			i := i + 1;
			j := j + 1;
			end
		else
			begin
			if (i >= n) or (krajnji[i] <> rezultat[k]) then
				begin
				writeln('Neispravna transformacija.');
				Exit;
				end;
			i := i + 1;
			end;
	end;
	if (i <> n) or (j <> m) then
		writeln('Neispravna transformacija.')
	else
		writeln('Ispravna transformacija u ', r ,' koraka.');
end.
